src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,836 @@
+                Pageable<TableLease> partitions = this.partitionTable.Query<TableLease>();
+                int ownershipLeaseCount = 0;
+                foreach (TableLease partition in partitions)
+                {
+                    bool DrainedLease = false;
+                    bool ReleasedLease = false;
+                    bool RenewedLease = false;
+
+                    if (partition.CurrentOwner == this.workerName)
+                    {
+                        ownershipLeaseCount++;
+                        ETag etag = partition.ETag;
+
+                        bool isDrainTaskExists = this.drainTask.TryGetValue(partition.RowKey!, out Task? task);
+                        if (isDrainTaskExists)
+                        {
+                            // Check if draining process has been finished.If so, release the lease.
+                            if (task.IsCompleted == true)
+                            {
+                                this.settings.Logger.PartitionManagerInfo(
+                                     this.storageAccountName,
+                                     this.settings.TaskHubName,
+                                     this.settings.WorkerId,
+                                     partition.RowKey,
+                                     $"Successfully drained the in-memory messages of lease {partition.RowKey} for shutdown. Lease will be released.");
+
+                                this.drainTask.Remove(partition.RowKey!);
+                                this.ReleaseLease(partition);
+                                ReleasedLease = true;
+                                ownershipLeaseCount--;
+                                await task; // Catch any exceptions happened in the drain process.
+                            }
+                            else// If draining process is not finished, keep renewing the lease in case it's expired during the draining.
+                            {
+                                this.RenewLease(partition);
+                                RenewedLease = true;
+                            }
+                        }
+                        else 
+                        {
+                            this.DrainLease(partition, CloseReason.Shutdown);
+                            this.RenewLease(partition);
+                            RenewedLease = true;
+                            DrainedLease = true;
+                        }
+                        
+                        try
+                        {
+                            await this.partitionTable.UpdateEntityAsync(partition, etag, (TableUpdateMode)1);
+                            this.LogHelper(partition, false, false, RenewedLease, DrainedLease, ReleasedLease, "");
+                            
+                        }
+                        // Catch exception of failed table update due to ETag expired. 
+                        // Re-read the table immediately to get the latest ETag.
```

Comment:
In here, we're just throwing the exception and re-reading the table happens elsewhere. So I think you can just say the following:

```suggestion
                // Exception Status 412 represents an out of date ETag.
```